stan/math/prim/mat/fun/acos.hpp

```diff
@@ -27,11 +29,23 @@ struct acos_fun {
  * @tparam T Container type.
  * @return Arc cosine of each variable in the container, in radians.
  */
-template <typename T>
-inline typename apply_scalar_unary<acos_fun, T>::return_t acos(const T& x) {
+template <typename T, typename = require_not_eigen_vt<std::is_arithmetic, T>>
+inline auto acos(const T& x) {
   return apply_scalar_unary<acos_fun, T>::apply(x);
 }
 
+/**
+ * Version of acos() that accepts Eigen Matrix ar matrix expressions.
+ * @tparam Derived derived type of x
+ * @param x Matrix or matrix expression
+ * @return Arc cosine of each variable in the container, in radians.
+ */
+template <typename Derived,
+          typename = require_eigen_vt<std::is_arithmetic, Derived>>
+inline auto acos(const Eigen::MatrixBase<Derived>& x) {
+  return x.derived().array().acos().matrix();
+}
+
```

Comment:
This implementation will call Eigen's ```acos``` for Eigen objects (```Matrix```,```Array```,etc.), but not ```std::vectors``` of Eigen objects - as those will be handled by ```apply_scalar_unary```, so there's some possible performance gain there